source/memcached-store.ts

```diff
@@ -0,0 +1,171 @@
+		} else
+			this.client = new Memcached(
+				options?.locations ?? ['localhost:11211'],
+				options?.config ?? {},
+			)
+	}
+
+	/**
+	 * Method that actually initializes the store.
+	 *
+	 * @param options {RateLimitConfiguration} - The options used to setup the middleware.
+	 *
+	 * @impl
+	 */
+	init(options: RateLimitConfiguration) {
+		this.expiration = options.windowMs / 1000 // [in seconds]
+	}
+
+	/**
+	 * Method to prefix the keys with the given text.
+	 *
+	 * @param key {string} - The key.
+	 *
+	 * @returns {string} - The text + the key.
+	 */
+	prefixKey(key: string): string {
+		return `${this.prefix}${key}`
+	}
+
+	/**
+	 * Method to suffix the keys with `__expiry`, which is the name of the key used
+	 * to store the reset timestamp for that key.
+	 *
+	 * @param key {string} - The key.
+	 *
+	 * @returns {string} - The key + '__expiry'.
+	 */
+	expirySuffix(key: string): string {
+		return `${key}__expiry`
+	}
+
+	/**
+	 * Method to increment a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client.
+	 *
+	 * @returns {IncrementResponse} - The number of hits and reset time for that client.
+	 */
+	async increment(key: string): Promise<IncrementResponse> {
+		const prefixedKey = this.prefixKey(key)
+		const getKey = promisify(this.client.get).bind(this.client)
+		const setKey = promisify(this.client.set).bind(this.client)
+		const incrementKey = promisify(this.client.incr).bind(this.client)
+
+		// Try incrementing the given key. If the key exists, it will increment it
+		// and return the updated hit count.
+		// @ts-expect-error `incrementKey` returns a number or a boolean, not void.
+		// eslint-disable-next-line @typescript-eslint/no-confusing-void-expression
+		let totalHits = (await incrementKey(prefixedKey, 1)) as number | boolean
+		let expiresAt
+
+		if (totalHits === false) {
+			// The increment command failed since the key does not exist. In which case, set the
+			// hit count for that key to 1, and make sure it expires after `window` seconds.
+			await setKey(prefixedKey, 1, this.expiration)
+			totalHits = 1 // When you set it to 1, it returns `true` for some reason.
+
+			// Also store the expiration time in a separate key.
+			expiresAt = Date.now() + this.expiration
+			await setKey(
+				this.expirySuffix(prefixedKey), // The name of the key.
+				expiresAt, // The value - the time at which the key expires.
+				this.expiration, // The key should be deleted by memcached after `window` seconds.
+			)
+		} else {
+			// If the key exists and has been incremented succesfully, retrieve its expiry.
+			expiresAt = (await getKey(this.expirySuffix(prefixedKey))) as number
+		}
+
+		if (typeof totalHits !== 'number')
+			throw new Error(
+				`Expected 'totalHits' to be a number, got ${totalHits} instead.`,
+			)
+
+		// Return the total number of hits, as well as the reset timestamp.
+		return {
+			totalHits,
+			resetTime: expiresAt === undefined ? undefined : new Date(expiresAt),
```

Comment:
I think undefined would only happen if it expired in between reading the hits and reading the expired time, right? So perhaps we could default it to "right now"?
```suggestion
			// if expiresAt is undefined, assume the key expired in between reading the hits and expiry keys from memcached
			resetTime: expiresAt ? new Date(expiresAt) : new Date(),
```

Reply:
Hmm makes sense